sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableDestinationCoder.java

```diff
@@ -45,13 +46,20 @@ public void encode(TableDestination value, OutputStream outStream)
     }
     tableSpecCoder.encode(value.getTableSpec(), outStream);
     tableDescriptionCoder.encode(value.getTableDescription(), outStream);
+    timePartitioningCoder.encode(value.getJsonTimePartitioning(), outStream);
   }
 
   @Override
   public TableDestination decode(InputStream inStream) throws IOException {
     String tableSpec = tableSpecCoder.decode(inStream);
     String tableDescription = tableDescriptionCoder.decode(inStream);
-    return new TableDestination(tableSpec, tableDescription);
+    String jsonTimePartitioning = null;
+    try {
+      jsonTimePartitioning = timePartitioningCoder.decode(inStream);
+    } catch (IOException e) {
+      // This implies we're decoding old state that did not contain TimePartitioning. Continue.
```

Comment:
Another possibility - StringUtf8Coder passes EOFException as the cause if there is nothing to read. We can simply make the existing check more precise - check for EOFException instead of all IOExceptions.